# Segmenting and Clustering Neighborhoods in Toronto

## 1.Data structure from wiki

In [1]:
from bs4 import BeautifulSoup
import urllib.request
import pandas as pd
import numpy as np

#### *Use BeautifulSoup to get the information from the website*

In [2]:
#Get the HTML file from the URL
html = urllib.request.urlopen('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').read()

In [3]:
soup = BeautifulSoup(html)

In [4]:
#Find the table in the HTML file and extract it into a dataframe
table = soup.find('table', attrs = {'class':'wikitable sortable'})
table_rows = table.find_all('tr')
res = []

for tr in table_rows:
    td = tr.find_all('td')
    row = [tr.text.replace("\n", "") for tr in td]
    if row:
        res.append(row)
        
df = pd.DataFrame(res, columns = ["Postal Code", "Borough", "Neighborhood"])
df

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [5]:
#Ignore cells with a borough that is Not assigned.
df = df[df.Borough != 'Not assigned']
df.shape

(103, 3)

In [6]:
df['Postal Code'].unique

<bound method Series.unique of 2      M3A
3      M4A
4      M5A
5      M6A
6      M7A
      ... 
160    M8X
165    M4Y
168    M7Y
169    M8Y
178    M8Z
Name: Postal Code, Length: 103, dtype: object>

#### *We have 103 rows in the dataframe (after the ignoration of Not assigned Borough).* 
#### *And we have 103 different postal code in the dataframe. See the unique() result.*
#### *Wiki may redo this table and the problem doesn't exist anymore.*

*Combine those neighborhoods who share the same postal caode into one row and separate them with a comma.*  
*If you want to run the combination for sure, uncomment the code below.*

In [7]:
#df.groupby('Postal Code').agg({'Borough' : 'first', 'Neighborhood' : ','.join}).reset_index().reindex(columns=df.columns)

In [8]:
#Find if the column of Neighborhood contain the string "Not assigned"
a = df.Neighborhood

if 'Not assigned' in a:
    print("The column of Neighborhood contains string 'Not assigned'")
else:
    print('The column of Neighborhood is good')

The column of Neighborhood is good


In [9]:
df = df.reset_index(drop = True)
print (df.shape)
df

(103, 3)


Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

## 2. Add location information to the Data Frame

In [10]:
#Read the csv file from url to the data frame
url = 'https://cocl.us/Geospatial_data'
df_L = pd.read_csv(url)
df_L

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [11]:
#Use postal code as the index
df_L.set_index('Postal Code', inplace = True)
df.set_index('Postal Code', inplace = True)

In [12]:
#Inner join these two dataframe 
df_with_location = pd.concat([df, df_L], axis = 1, join = 'inner')
df_with_location.sort_values(by = 'Postal Code', axis = 0, inplace = True)
df_with_location.reset_index(inplace = True)
df_with_location

Postal Code      Borough  \
0           M1B  Scarborough   
1           M1C  Scarborough   
2           M1E  Scarborough   
3           M1G  Scarborough   
4           M1H  Scarborough   
..          ...          ...   
98          M9N         York   
99          M9P    Etobicoke   
100         M9R    Etobicoke   
101         M9V    Etobicoke   
102         M9W    Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                       Malvern, Rouge  43.806686 -79.194353  
1               Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Kingsview Village, St. Phillips, Martin Grove ...  43.688905 -79.554724  
101  South Steeles, Silverstone, Humbergate, Jamest...  43.739416 -79.588437  
102                Northwest, West Humber - Clairville  43.706748 -79.594054  

[103 rows x 5 columns]

## 3.Segmenting_and_Clustering_Neighborhoods_in_Toronto

In [13]:
#!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim
#!conda install -c conda-forge folium=0.5.0 --yes
import folium
import requests
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors
import os

In [14]:
#Get toronto data
toronto_data = df_with_location.loc[df_with_location['Borough'].isin(
    ['Central Toronto', 
     'Downtown Toronto', 
     'East Toronto'
    ])].reset_index(drop = True)

toronto_data.head()

Postal Code          Borough                    Neighborhood   Latitude  \
0         M4E     East Toronto                     The Beaches  43.676357   
1         M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2         M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3         M4M     East Toronto                 Studio District  43.659526   
4         M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [15]:
toronto_data.shape

(33, 5)

In [16]:
#Get the latitude and the longitude of Toronto
address = 'Toronto, CA'

geolocator = Nominatim(user_agent = "To_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [17]:
#Create a map of Toronto using latitude and longitude values
map_Toronto = folium.Map(location = [latitude, longitude], zoom_start = 11)

#Add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
        [lat, lng],
        radius = 5,
        popup = label,
        color = 'blue',
        fill = True,
        fill_color = '#3186cc',
        fill_opacity = 0.7,
        parse_html = False).add_to(map_Toronto)  
    
map_Toronto

In [19]:
#Use the Foursquare API to explore the neighborhoods and segment them
CLIENT_ID = os.environ.get('Foursquare_ID')
CLIENT_SECRET = os.environ.get('Foursquare_SECRET')
VERSION = '20200605' 

LIMIT = 100
radius = 500

**Explore Neighborhoods in Toronto**

In [20]:
def getNearbyVenues(names, latitudes, longitudes, radius = 500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [21]:
#Run the getNearbyVenues fonction and create a new dataframe called toronto_venues
toronto_venues = getNearbyVenues(names = toronto_data['Neighborhood'],
                                   latitudes = toronto_data['Latitude'],
                                   longitudes = toronto_data['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West,  Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Queen's Park, Ontario Provincial Government
Business reply mail Processing Centre, South Central Letter

In [22]:
#Check the size of the resulting dataframe
print(toronto_venues.shape)
toronto_venues.head()

(1464, 7)


Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

In [23]:
#Check how many venues were returned for each neighborhood
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
Business reply mail Processing Centre, South Ce...                     17   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     65   
Christie                                                               16   
Church and Wellesley                                                   77   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        8   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
India Bazaar, The Beaches West                                         24   
Kensington Market, Chinatown, Grange Park                              59   
Lawrence Park                                                           3   
Moore Park, Summerhill East                                             2   
North Toronto West,  Lawrence Park                                     19   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
Roselawn                                                                2   
St. James Town                                                         80   
St. James Town, Cabbagetown                                            44   
Stn A PO Boxes                                                         97   
Studio District                                                        40   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    23   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
Business reply mail Processing Centre, South Ce...                      17   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      65   
Christie                                                                16   
Church and Wellesley                                                    77   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              35   
Davisville North                                                         8   
First Canadian Place, Underground city                                 100   
Forest Hill North & West, Forest Hill Road Park                          4   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
India Bazaar, The Beaches West                                          24   
Kensington Market, Chinatown, Grange Park                               59   
Lawrence Park                 

In [24]:
#Check how many unique categories can be curated from all the returned venues
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 224 uniques categories.


In [25]:
#One hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix = "", prefix_sep = "")

#Add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

#Move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Theme Restaurant  Toy / Game Store  Trail  \
0             0  ...                 0                 0      1   
1             0  ...                 0                 0      0   
2             0  ...                 0                 0      0   
3             0  ...                 0                 0      0   
4             0  ...                 0                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  
0                      0         0          0              0  
1                      0         0          0              0  
2                      0         0          0              0  
3                      0         0          0              0  
4                      0         0          0              0  

[5 rows x 224 columns]

In [26]:
#Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()

In [27]:
#Sort the venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending = False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

**create the new dataframe and display the top 10 venues for each neighborhood.**

In [28]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns = columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  Business reply mail Processing Centre, South C...           Yoga Studio   
2  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar              Beer Bar                Bakery   
1         Auto Workshop                  Park            Comic Shop   
2        Airport Lounge      Airport Terminal           Coffee Shop   
3    Italian Restaurant        Sandwich Place   Japanese Restaurant   
4                  Café                  Park             Nightclub   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0    Seafood Restaurant            Restaurant           Cheese Shop   
1           Pizza Place      Recording Studio            Restaurant   
2       Harbor / Marina                 Plane   Rental Car Location   
3                  Café       Bubble Tea Shop           Salad Place   
4                 Diner    Italian Restaurant            Baby Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Café                   Pub          Shopping Mall  
1               Butcher         Burrito Place                Brewery  
2      Sculpture Garden              Boutique          Boat or Ferry  
3       Thai Restaurant      Department Store                    Bar  
4           Candy Store            Restaurant            Coffee Shop

**Run k-means to cluster the neighborhood into 5 clusters.**

In [29]:
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

#Run k-means clustering
kmeans = KMeans(n_clusters = kclusters, random_state = 0).fit(toronto_grouped_clustering)

#Check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 0, 2, 2, 2, 2, 2], dtype=int32)

In [30]:
#Add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on = 'Neighborhood')

toronto_merged.head()

Postal Code          Borough                    Neighborhood   Latitude  \
0         M4E     East Toronto                     The Beaches  43.676357   
1         M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2         M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3         M4M     East Toronto                 Studio District  43.659526   
4         M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               2                 Trail     Health Food Store   
1 -79.352188               2      Greek Restaurant           Coffee Shop   
2 -79.315572               2                  Park        Sandwich Place   
3 -79.340923               2                  Café           Coffee Shop   
4 -79.388790               0              Bus Line                  Park   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0                   Pub         Women's Store            Dance Studio   
1    Italian Restaurant            Restaurant  Furniture / Home Store   
2  Fast Food Restaurant           Pizza Place       Food & Drink Shop   
3             Gastropub                Bakery                 Brewery   
4           Swim School         Women's Store       Electronics Store   

         6th Most Common Venue        7th Most Common Venue  \
0            Electronics Store  Eastern European Restaurant   
1                    Bookstore               Ice Cream Shop   
2                      Brewery                Burrito Place   
3          American Restaurant                  Yoga Studio   
4  Eastern European Restaurant                   Donut Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Donut Shop      Doner Restaurant                Dog Run  
1                   Pub           Pizza Place                 Lounge  
2            Restaurant    Italian Restaurant         Ice Cream Shop  
3    Seafood Restaurant        Sandwich Place            Cheese Shop  
4      Doner Restaurant               Dog Run    Distribution Center

**Create map**

In [31]:
map_clusters = folium.Map(location = [latitude, longitude], zoom_start = 11)

#Set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

#Add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html = True)
    folium.CircleMarker(
        [lat, lon],
        radius = 5,
        popup = label,
        color = rainbow[cluster-1],
        fill = True,
        fill_color = rainbow[cluster-1],
        fill_opacity = 0.7).add_to(map_clusters)
       
map_clusters

Cluster 1

In [32]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4    Central Toronto               0              Bus Line   
30  Downtown Toronto               0         Grocery Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4                   Park           Swim School         Women's Store   
30                  Café                  Park             Nightclub   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
4      Electronics Store  Eastern European Restaurant            Donut Shop   
30                 Diner           Italian Restaurant            Baby Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4       Doner Restaurant               Dog Run    Distribution Center  
30           Candy Store            Restaurant            Coffee Shop

Cluster 2

In [33]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               1                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22           Music Venue         Women's Store          Dance Studio   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
22     Electronics Store  Eastern European Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22      Doner Restaurant               Dog Run    Distribution Center

Cluster 3

In [34]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               2                 Trail   
1       East Toronto               2      Greek Restaurant   
2       East Toronto               2                  Park   
3       East Toronto               2                  Café   
5    Central Toronto               2                  Park   
6    Central Toronto               2        Clothing Store   
7    Central Toronto               2           Pizza Place   
9    Central Toronto               2                   Pub   
11  Downtown Toronto               2           Coffee Shop   
12  Downtown Toronto               2           Coffee Shop   
13  Downtown Toronto               2           Coffee Shop   
14  Downtown Toronto               2        Clothing Store   
15  Downtown Toronto               2           Coffee Shop   
16  Downtown Toronto               2           Coffee Shop   
17  Downtown Toronto               2           Coffee Shop   
18  Downtown Toronto               2           Coffee Shop   
19  Downtown Toronto               2           Coffee Shop   
20  Downtown Toronto               2           Coffee Shop   
21  Downtown Toronto               2           Coffee Shop   
24   Central Toronto               2           Coffee Shop   
25  Downtown Toronto               2                  Café   
26  Downtown Toronto               2                  Café   
27  Downtown Toronto               2       Airport Service   
28  Downtown Toronto               2           Coffee Shop   
29  Downtown Toronto               2           Coffee Shop   
31  Downtown Toronto               2           Coffee Shop   
32      East Toronto               2           Yoga Studio   

   2nd Most Common Venue  3rd Most Common Venue  \
0      Health Food Store                    Pub   
1            Coffee Shop     Italian Restaurant   
2         Sandwich Place   Fast Food Restaurant   
3            Coffee Shop              Gastropub   
5                    Gym         Breakfast Spot   
6            Coffee Shop          Grocery Store   
7           Dessert Shop         Sandwich Place   
9            Coffee Shop            Pizza Place   
11           Pizza Place             Restaurant   
12      Sushi Restaurant    Japanese Restaurant   
13                Bakery                    Pub   
14           Coffee Shop                   Café   
15                  Café    American Restaurant   
16          Cocktail Bar               Beer Bar   
17    Italian Restaurant         Sandwich Place   
18                  Café             Restaurant   
19              Aquarium                   Café   
20                 Hotel                   Café   
21            Restaurant                   Café   
24                  Café         Sandwich Place   
25            Restaurant                    Bar   
26                Bakery  Vietnamese Restaurant   
27        Airport Lounge       Airport Terminal   
28    Seafood Restaurant                   Café   
29                  Café             Restaurant   
31      Sushi Restaurant                  Diner   
32         Auto Workshop                   Park   

            4th Most Common Venue   5th Most Common Venue  \
0                   Women's Store            Dance Studio   
1                      Restaurant  Furniture / Home Store   
2                     Pizza Place       Food & Drink Shop   
3                          Bakery                 Brewery   
5                         Dog Run       Food & Drink Shop   
6             Sporting Goods Shop                    Café   
7                Sushi Restaurant             Coffee Shop   
9                Sushi Restaurant              Bagel Shop   
11                           Café      Italian Restaurant   
12                     Restaurant                 Gay Bar   
13                           Park          Breakfast Spot   
14                Bubble Tea Shop     Japanese Restaurant   
15                      Gastropub              Restaur

Cluster 4

In [35]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
10  Downtown Toronto               3                  Park   
23   Central Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10            Playground                 Trail          Dance Studio   
23         Jewelry Store                 Trail      Sushi Restaurant   

   5th Most Common Venue        6th Most Common Venue  \
10     Electronics Store  Eastern European Restaurant   
23         Deli / Bodega            Electronics Store   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
10                   Donut Shop      Doner Restaurant               Dog Run   
23  Eastern European Restaurant            Donut Shop      Doner Restaurant   

   10th Most Common Venue  
10    Distribution Center  
23                Dog Run

Cluster 5

In [36]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
8  Central Toronto               4                   Gym   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8                 Trail         Women's Store         Deli / Bodega   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
8  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
8            Donut Shop      Doner Restaurant                Dog Run